### PR TITLE
feat (types): add complete `fastify.listen()` typescript definitions

### PR DESCRIPTION
**Pull request by darkgl0w**
Hello.

This PR aims to :
- add complete `fastify.listen()` typescript definitions (cf.: https://github.com/fastify/fastify/pull/3712#issuecomment-1058015854)
- add deprecations on typescript side to promote the use of the object options

#### Checklist

- [x] run `npm run test` ~~and `npm run benchmark`~~
- [x] tests ~~and/or benchmarks~~ are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: types/instance.d.ts [hunk 1]
```diff
@@ -82,12 +82,117 @@ export interface FastifyInstance<
...
+   */
   listen(port: number | string, callback: (err: Error|null, address: string) => void): void;
+  /**
+   * @deprecated Variadic listen method is deprecated. Please use `.listen(optionsObject)` instead. The variadic signature will be removed in `fastify@5`
+   * @see https://github.com/fastify/fastify/pull/3712
+   */
   listen(port: number | string, address?: string, backlog?: number): Promise<string>;
-  listen(opts: { port: number; host?: string; backlog?: number }, callback: (err: Error|null, address: string) => void): void;
-  listen(opts: { port: number; host?: string; backlog?: number }): Promise<string>;
+  listen(callback: (err: Error | null, address: string) => void): void;
```
Comment (darkgl0w) [inline types/instance.d.ts:105]: `fastify.listen(() => {})` call is valid, it falls back to listening on
```js
{
  host: 'localhost', // both ipv4 and ipv6
  port: 0
}
```

File: types/instance.d.ts [hunk 2]
```diff
@@ -82,12 +82,117 @@ export interface FastifyInstance<
...
+     * Default to `false`.
+     */
+    ipv6Only?: boolean;
+    /**
+     * An AbortSignal that may be used to close a listening server.
+     * @since This option is available only in Node.js v15.6.0 and greater
+     */
+    signal?: AbortSignal;
+  }, callback: (err: Error|null, address: string) => void): void;
+  listen(opts?: {
```
Comment (darkgl0w) [inline types/instance.d.ts:151]: `fastify.listen()` call is valid, it falls back to listening on
```js
{
  host: 'localhost', // both ipv4 and ipv6
  port: 0
}
```

File: types/instance.d.ts [hunk 3]
```diff
@@ -82,12 +82,117 @@ export interface FastifyInstance<
   inject(opts: InjectOptions | string): Promise<LightMyRequestResponse>;
   inject(): LightMyRequestChain;
 
+  /**
+   * @deprecated Variadic listen method is deprecated. Please use `.listen(optionsObject)` instead. The variadic signature will be removed in `fastify@5`
+   * @see https://github.com/fastify/fastify/pull/3712
```
Comment (darkgl0w) [inline types/instance.d.ts:87]: Suggestions are welcome.
For now I choose to display the same message as the warning and to link the GitHub PR acting this change.

**Review by darkgl0w — COMMENTED** [review]
This PR is ready to be reviewed.

**Review by RafaelGSS — APPROVED** [review]
LGTM.

**Pull request closed without merging**